### PR TITLE
refactor: drop usage of chainstate globals in Dash-specific code, merge bitcoin#21866 (goodbye to a global chainstate)

### PR DESCRIPTION
**Pull request by kwvg**
```
Thank you, I'll say goodbye soon
Though its the end of these globals, don't blame yourself now
And if its true, I will surround you and give life to a chainstate
That's our own
```

## Additional Information

* In `CDeterministicMN::ToJson()`, `collateralAddress` is extracted by finding the `scriptPubKey` of a transaction output for a masternode, originally this used `GetUTXOCoin` but doesn't work for spent tranasction outputs (as they're _not_ UTXOs), so in [dash#5607](https://github.com/dashpay/dash/pull/5607), a fallback was introduced that looks through the general transaction set if going through the UTXO set yielded nothing.

   `GetUTXOCoin` accesses the active chainstate to get ahold of the UTXO set, this was done through globals. The removal of chainstate globals meant that whoever was calling `GetUTXOCoin` should have access to the chainstate handy. This is trivial in RPC code where `ToJson()` is used ([source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/rpc/evo.cpp#L1286)) through `Ensure`(`Any`)`Chainman`. Not the case in Qt code ([source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/qt/masternodelist.cpp#L369)), which is supposed to be given restricted access to information by the interface.

  As the fallback seems to be capable of fetching UTXOs and spent outputs, we can remove the `GetUTXOCoin` method and make the fallback the only method.

* In `develop`, as of this writing, `CChainState` members `FlushStateToDisk` and {`Enforce`, `Invalidate`, `MarkConflicting`}`Block` were accessing their internals through the global, despite having direct access to them. As the globals they were calling are going to be bid farewell, they needed to be changed to access its members instead.

  The reason for going the roundabout way is unknown.

* `CDSNotificationInterface` takes in a `ChainstateManager` (instead of the `CChainState` it actually requires) as at the time of interface initialization ([source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/init.cpp#L1915-L1918)), the active chainstate hasn't been loaded in yet as that happens further down ([source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/init.cpp#L1988-L1991)).

  As `CDSNotificationInterface::InitializeCurrentBlockTip()` is called well after it is initialized, we can resolve to the active chainstate in there.

* As `GetCreditPoolDiffForBlock` requires access to `ChainstateManager` as `GetCreditPoolDiffForBlock` > `ProcessLockUnlockTransaction` > `CheckAssetLockUnlockTx` > `CheckAssetUnlockTx` > `ChainstateManager::m_blockman.LookupBlockIndex()` and `BlockAssembler` only has `CChainState`, it had to be reworked around `ChainstateManager`.

  ~~`CChainState` is passed as a direct argument while `ChainstateManager` can be fetched from `NodeContext`. Unlike `CTxMemPool`, which can be passed custom instances ([source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/rpc/mining.cpp#L381-L382), [source](https://github.com/dashpay/dash/blob/5baa5222258f5cf2be0a3ce0f335dfd2fee931bc/src/test/util/setup_common.cpp#L391-L392)), `CChainState`'s argument value is taken from `NodeContext::chainstate.ActiveChainstate()` and since we're now accepting `ChainstateManager` wholesale, we can dispense with accepting `CChainState` as an argument.~~

  ~~Changes to that effect have been made.~~

  AssumeUTXO introduces the need to be able to use different `CChainState`s, so this underlying assumption no longer holds true, the above described changes have been reverted. Asset locks code has been refactored to use `BlockManager` directly (which does come with the downside of needing to hold `cs_main` for longer than strictly necessary, this is why only asset locks uses `BlockManager` directly while other cases still benefit from having `ChainstateManager` as a whole).

* `CMNHFManager::ConnectManagers` will be taking in a `ChainstateManager` pointer due to the `GetSignalsStage` > `GetForBlock` > `ProcessBlock` > `extractSignals` > `CheckMNHFTx` > `ChainstateManager::m_blockman.LookupBlockIndex()` chain.

* The use of a bespoke `NodeContext` in `coinselector_tests` breaks tests if any interface call relies on a chainstate as `testNode` doesn't initialize one. For the most part, this was masked by `WalletTestingSetup` populating the chainstate globals from its own `NodeContext` even if the tests themselves preferred to use their own stripped down `testNode`. 

  Though, removing the chainstate globals meant that they can no longer rely on `WalletTestingSetup`'s `NodeContext` to mask the barebones `testNode` global being used in the test (specifically, `addCoins` > `listMNCollaterials` > `ChainActive()` worked because `ChainActive()` accessed `WalletTestingSetup`'s `NodeContext` but when `ChainActive()` was gone and replaced with `NodeContext::chainman.ActiveChain()`, it uses `testNode`'s `ChainstateManager`, which doesn't exist, which causes it to crash).

  To remedy this, a5595b13 and 5e54aa9b from [bitcoin#23288](https://github.com/bitcoin/bitcoin/pull/23288) were adapted for the limited purpose of eliminating `testNode` and using `WalletTestingSetup`'s `NodeContext` instead. This comes with the unfortunate effect of skipping a lot of the refactoring, cleanups and optimizations done before and adapting the ones after them non-trivial.

  It is therefore best recommended that the commit be reverted and changes implemented step-by-step in a pull request at some point in the future. For now, it's kept around here for the sake of this pull request, which, if merged, should prevent more chainstate globals use from leaking into the codebase.

    <details>

    <summary>Pre-fix crash stacktrace: </summary>
    
    ```
    dash@71aecd6afb45:/src/dash$ lldb-16 ./src/test/test_dash
    (lldb) target create "./src/test/test_dash"
    Current executable set to '/src/dash/src/test/test_dash' (x86_64).
    (lldb) r -t coinselector_tests
    Process 395006 launched: '/src/dash/src/test/test_dash' (x86_64)
    Running 4 test cases...
    node/interfaces.cpp:711 chainman: Assertion `m_node.chainman' failed.
    Process 395006 stopped
    * thread #1, name = 'd-test', stop reason = signal SIGABRT
        frame #0: 0x00007ffff7a7300b libc.so.6`__GI_raise(sig=<unavailable>) at raise.c:51:1
    (lldb) bt
    * thread #1, name = 'd-test', stop reason = signal SIGABRT
    * frame #0: 0x00007ffff7a7300b libc.so.6`__GI_raise(sig=<unavailable>) at raise.c:51:1
        frame #1: 0x00007ffff7a52859 libc.so.6`__GI_abort at abort.c:79:7
        frame #2: 0x00005555563cba33 test_dash`assertion_fail(file="node/interfaces.cpp", line=711, func="chainman", assertion="m_node.chainman") at check.cpp:13:5
        frame #3: 0x0000555555fb47aa test_dash`node::(anonymous namespace)::ChainImpl::listMNCollaterials(std::vector<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>, std::allocator<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>>> const&) [inlined] std::unique_ptr<ChainstateManager, std::default_delete<ChainstateManager>>& inline_assertion_check<true, std::unique_ptr<ChainstateManager, std::default_delete<ChainstateManager>>&>(val=nullptr, file=<unavailable>, line=711, func=<unavailable>, assertion=<unavailable>) at check.h:62:13
        frame #4: 0x0000555555fb4781 test_dash`node::(anonymous namespace)::ChainImpl::listMNCollaterials(std::vector<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>, std::allocator<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>>> const&) [inlined] node::(anonymous namespace)::ChainImpl::chainman(this=0x000055555723e830)at interfaces.cpp:711:45
        frame #5: 0x0000555555fb477d test_dash`node::(anonymous namespace)::ChainImpl::listMNCollaterials(std::vector<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>, std::allocator<std::pair<std::shared_ptr<CTransaction const> const&, unsigned int>>> const&) [inlined] node::(anonymous namespace)::ChainImpl::listMNCollaterials(this=<unavailable>)::'lambda'()::operator()() const at interfaces.cpp:788:34
        frame #6: 0x0000555555fb474f test_dash`node::(anonymous namespace)::ChainImpl::listMNCollaterials(this=0x000055555723e830, outputs=size=0) at interfaces.cpp:788:34
        frame #7: 0x00005555565bcd07 test_dash`CWallet::AddToWallet(this=0x00005555571701e0, tx=<unavailable>, confirm=<unavailable>, update_wtx=<unavailable>, fFlushOnClose=<unavailable>) at wallet.cpp:886:46
        frame #8: 0x0000555555bed3ef test_dash`coinselector_tests::add_coin(wallet=0x00005555571701e0, nValue=0x00007fffffffc7c0, nAge=144, fIsFromMe=false, nInput=0, spendable=<unavailable>) at coinselector_tests.cpp:77:29
        frame #9: 0x0000555555bead3e test_dash`coinselector_tests::bnb_search_test::test_method() [inlined] coinselector_tests::add_coin(nValue=0x00007fffffffc7c0, nAge=144, fIsFromMe=false, nInput=0, spendable=false) at coinselector_tests.cpp:88:5
        frame #10: 0x0000555555bead20 test_dash`coinselector_tests::bnb_search_test::test_method(this=0x00007fffffffcad0) at coinselector_tests.cpp:278:5
        frame #11: 0x0000555555be6607 test_dash`coinselector_tests::bnb_search_test_invoker() at coinselector_tests.cpp:138:1
    ```

    </details>

## Breaking Changes

* Backporting `coinselector_tests` changes are now much more annoying.

* The following RPCs, `protx list`, `protx listdiff`, `protx info` will no longer report `collateralAddress` if the transaction index has been disabled (`txindex=0`).

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas **(note: N/A)**
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation **(note: N/A)**
- [x] I have assigned this pull request to a milestone _(for repository code-owners and collaborators only)_

### CODE REVIEW

File: src/evo/deterministicmns.cpp [hunk 1]
```diff
@@ -49,17 +49,13 @@ UniValue CDeterministicMN::ToJson() const
...
-    CTxDestination dest;
-    if (ExtractDestination(scriptPubKey, dest)) {
-        obj.pushKV("collateralAddress", EncodeDestination(dest));
+    uint256 tmpHashBlock;
+    CTransactionRef collateralTx = GetTransaction(/* block_index */ nullptr,  /* mempool */ nullptr, collateralOutpoint.hash, Params().GetConsensus(), tmpHashBlock);
+    if (collateralTx) {
+        CTxDestination dest;
+        if (ExtractDestination(collateralTx->vout[collateralOutpoint.n].scriptPubKey, dest)) {
+            obj.pushKV("collateralAddress", EncodeDestination(dest));
+        }
```
Comment (PastaPastaPasta) [inline src/evo/deterministicmns.cpp:58]: needs release notes documenting breaking change when txindex is disabled

File: src/governance/classes.h [hunk 2]
```diff
@@ -38,7 +38,7 @@ class CSuperblockManager
     static bool GetSuperblockPayments(CGovernanceManager& govman, const CDeterministicMNList& tip_mn_list, int nBlockHeight, std::vector<CTxOut>& voutSuperblockRet);
     static void ExecuteBestSuperblock(CGovernanceManager& govman, const CDeterministicMNList& tip_mn_list, int nBlockHeight);
 
-    static bool IsValid(CGovernanceManager& govman, const CDeterministicMNList& tip_mn_list, const CTransaction& txNew, int nBlockHeight, CAmount blockReward);
+    static bool IsValid(CGovernanceManager& govman, const CChain &active_chain, const CDeterministicMNList& tip_mn_list, const CTransaction& txNew, int nBlockHeight, CAmount blockReward);
```
Comment (PastaPastaPasta) [inline src/governance/classes.h:41]: `const CChain& active_chain`

**Review by PastaPastaPasta — COMMENTED** [review]
general utACK; a few problems: 8e32493abceb8c22caa8abcff893101a7d061a94

**Review by knst — APPROVED** [review]
LGTM overall, but see comments

File: src/spork.cpp [hunk 3]
```diff
@@ -149,8 +150,8 @@ PeerMsgRet CSporkManager::ProcessSpork(const CNode& peer, PeerManager& peerman,
     {
         LOCK(cs_main);
         EraseObjectRequest(peer.GetId(), CInv(MSG_SPORK, hash));
-        if (!::ChainActive().Tip()) return {};
-        strLogMsg = strprintf("SPORK -- hash: %s id: %d value: %10d bestHeight: %d peer=%d", hash.ToString(), spork.nSporkID, spork.nValue, ::ChainActive().Height(), peer.GetId());
+        if (!active_chain.Tip()) return {};
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/spork.cpp:153]: Examining that condition is outside the scope of this PR. If that condition can be done away with (in a different PR before or after this one), the only remaining usage would be to print height, which can be trimmed away by omitting `bestHeight`, which will remove the need for passing `active_chain`.

File: src/miner.h [hunk 4]
```diff
@@ -159,18 +160,19 @@ class BlockAssembler
...
     CChainstateHelper& m_chain_helper;
+    CEvoDB& m_evoDb;
+    ChainstateManager& m_chainman;
     CMNHFManager& m_mnhfman;
-    const llmq::CQuorumBlockProcessor& m_quorum_block_processor;
-    const llmq::CQuorumManager& m_qman;
     llmq::CChainLocksHandler& m_clhandler;
     llmq::CInstantSendManager& m_isman;
-    CEvoDB& m_evoDb;
+    const CChainParams& chainparams;
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/miner.h:172]: `CChainParams` is replaced by `ChainstateManager` due to `GetCreditPoolDiffForBlock` need for the latter, introducing that is an opportunity to revisit the argument list and its ordering.

File: src/spork.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/spork.cpp:153]: Yeah, this was introduced back then when we used `chainActive.Tip()->nHeight` in logs and was never reviewed/removed when we switched to `chainActive.Height()` later. Should be safe to remove heigh/active_chain from spork logs imo. Could be done in a separate PR but technically speaking 68e677cdc292a04c235c77a25f1a396d225105f7 wasn't a refactor either so either way is fine imo.

File: src/validation.h [hunk 5]
```diff
@@ -1146,7 +1146,7 @@ extern std::unique_ptr<CBlockTreeDB> pblocktree;
  * Return true if hash can be found in ::ChainActive() at nBlockHeight height.
  * Fills hashRet with found hash, if no nBlockHeight is specified - ::ChainActive().Height() is used.
  */
-bool GetBlockHash(uint256& hashRet, int nBlockHeight = -1);
+bool GetBlockHash(const CChain& active_chain, uint256& hashRet, int nBlockHeight = -1);
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/validation.h:1142]: Resolved in latest push.

File: src/bench/coin_selection.cpp [hunk 6]
```diff
@@ -61,21 +61,16 @@ static void CoinSelection(benchmark::Bench& bench)
 }
 
 typedef std::set<CInputCoin> CoinSet;
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/bench/coin_selection.cpp:63]: Resolved in latest push.

File: src/spork.h [hunk 7]
```diff
@@ -258,15 +259,17 @@ class CSporkManager : public SporkStore
...
+                              std::string_view msg_type, CDataStream& vRecv);
 
     /**
      * ProcessSpork is used to handle the 'spork' p2p message.
      *
      * For 'spork', it validates the spork and adds it to the internal spork storage and
      * performs any necessary processing.
      */
-    PeerMsgRet ProcessSpork(const CNode& peer, PeerManager& peerman, CDataStream& vRecv) EXCLUSIVE_LOCKS_REQUIRED(!cs);
+    PeerMsgRet ProcessSpork(const CNode& peer, PeerManager& peerman, const CChain& active_chain, CDataStream& vRecv)
```
[the remark this answers is not in the document]
Comment (kwvg) [reply src/spork.h]: Resolved in latest push.

File: src/governance/classes.h (hunk 2 above)
Comment (kwvg) [reply src/governance/classes.h]: Resolved in latest push.

File: src/evo/deterministicmns.cpp (hunk 1 above)
Comment (kwvg) [reply src/evo/deterministicmns.cpp:58]: The following RPC calls are affected:

* `protx info` ([source](https://github.com/dashpay/dash/blob/4a520991db12ff1276aac0ea43a77fe9d400d52d/src/rpc/evo.cpp#L1469))
* `protx list` ([source](https://github.com/dashpay/dash/blob/4a520991db12ff1276aac0ea43a77fe9d400d52d/src/rpc/evo.cpp#L1376), [source](https://github.com/dashpay/dash/blob/4a520991db12ff1276aac0ea43a77fe9d400d52d/src/rpc/evo.cpp#L1402))
* `protx listdiff` ([source](https://github.com/dashpay/dash/blob/4a520991db12ff1276aac0ea43a77fe9d400d52d/src/rpc/evo.cpp#L1581-L1583))

Resolved in latest push.

**Comment by kwvg** [comment]
The latest diff of changes got buried due to a force push to fix a typo, the diff is available [here](https://github.com/dashpay/dash/compare/8e32493a..4646eb70).

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by knst — APPROVED** [review]
utACK 0213fbebe6b2ecde2c7217af6adfe114fb55058b

force-pushes looks good for me

**Review by PastaPastaPasta — APPROVED** [review]
utACK https://github.com/dashpay/dash/commit/0213fbebe6b2ecde2c7217af6adfe114fb55058b

File: src/wallet/test/coinselector_tests.cpp [hunk 8]
```diff
@@ -27,16 +27,14 @@ BOOST_FIXTURE_TEST_SUITE(coinselector_tests, WalletTestingSetup)
...
+/* TODO: Revert current globals removal and backport pull requests implementing it */
+static const CoinEligibilityFilter filter_standard(1, 6, 0);
+static const CoinEligibilityFilter filter_confirmed(1, 1, 0);
+static const CoinEligibilityFilter filter_standard_extra(6, 6, 0);
 
-CoinEligibilityFilter filter_standard(1, 6, 0);
-CoinEligibilityFilter filter_confirmed(1, 1, 0);
-CoinEligibilityFilter filter_standard_extra(6, 6, 0);
-CoinSelectionParams coin_selection_params(false, 0, 0, CFeeRate(0), 0, false);
+CoinSelectionParams coin_selection_params(/* use_bnb = */ false, /* change_output_size = */ 0,
```
Comment (knst) [inline src/wallet/test/coinselector_tests.cpp:35]: unrelated refactoring!
